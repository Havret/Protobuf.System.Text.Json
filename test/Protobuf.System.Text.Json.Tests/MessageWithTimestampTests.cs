using System.Text.Json.Protobuf.Tests.Utils;
using Google.Protobuf.WellKnownTypes;
using Shouldly;
using SmartAnalyzers.ApprovalTestsExtensions;
using Xunit;

namespace System.Text.Json.Protobuf.Tests;

public class MessageWithTimestampTests
{
    [Fact]
    public void Should_serialize_message_with_timestamp()
    {
        // Arrange
        var msg = new MessageWithTimestamp
        {
            TimestampProperty = Timestamp.FromDateTime(new DateTime(1991, 01, 25, 0, 0, 0, DateTimeKind.Utc))
        };
        
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_serialize_message_with_timestamp_when_value_not_set()
    {
        // Arrange
        var msg = new MessageWithTimestamp();
        
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_serialize_and_deserialize_message_with_timestamp()
    {
        // Arrange
        var msg = new MessageWithTimestamp
        {
            TimestampProperty = Timestamp.FromDateTime(new DateTime(1991, 01, 25, 0, 0, 0, DateTimeKind.Utc))
        };
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithTimestamp>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.ShouldBeEquivalentTo(msg);
    }
    
    [Fact]
    public void Should_serialize_and_deserialize_message_with_timestamp_when_value_is_not_set()
    {
        // Arrange
        var msg = new MessageWithTimestamp();
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithTimestamp>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.ShouldBeEquivalentTo(msg);
    }
}