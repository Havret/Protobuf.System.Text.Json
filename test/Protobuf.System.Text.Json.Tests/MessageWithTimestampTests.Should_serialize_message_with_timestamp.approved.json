{
  "timestampProperty": "1991-01-25T00:00:00Z"
}